File: leetcode_python/Recursion/binary-tree-longest-consecutive-sequence.py
# https://baihuqian.github.io/2018-08-01-binary-tree-longest-consecutive-sequence/
# https://leetcode.ca/all/298.html

"""

Leetcode 298: Binary Tree Longest Consecutive Sequence

Question
Given a binary tree, find the length of the longest consecutive sequence path.

The path refers to any sequence of nodes from some starting node to any node in the tree along the parent-child connections. The longest consecutive path need to be from parent to child (cannot be the reverse).

Example 1:

Input:

   1
    \
     3
    / \
   2   4
        \
         5

Output: 3

Explanation: Longest consecutive sequence path is 3-4-5, so return 3.
Example 2:

Input:

   2
    \
     3
    /
   2    
  /
 1

Output: 2

Explanation: Longest consecutive sequence path is 2-3, not 3-2-1, so return 2.

"""


"""
NOTE !!! below is WRONG



Reason:

1. Tree vs. Linear Path: 
   A consecutive sequence path only flows 
   from parent to child. Your serialization string
   "{}-{}-{}" combines a node with both its left and right subtrees. 
   This destroys the linear, top-to-bottom path 
   you actually need to check.


2. Syntax Errors: 
   self.p_map().keys() 
   will throw an error because a dictionary is not callable. 
   Also, len = 0 overwrites Python's built-in len() function,
    causing len(_list) to fail.

3. Performance: 
   String manipulation and splitting take 
   significantly more time and space than 
   simply traversing the tree.


(WRONG !!!)

```
class Solution(object):
    def longestConsecutive(self, root):
        # edge
        if not root:
            return 0
        if not root.left and not root.right:
            return 1

        self.max_len = 1
        # {path: cnt}
        self.p_map = {}

        self.helper(root)

        for p in self.p_map().keys():
            self.max_len = max(self.max_len, self.get_len(p))


        return self.max_len


     def get_len(self, path):
        if not path or path == "#":
            return 0

        _list = path.split("-")
        len = 0

        if len(_list) == 1:
            return 1

        start = 0
        for i in range(1, len(_list) - 1):
            if int(_list[i]) != int(_list[i-1]) + 1:
                return 0

        return len(_list) - 1


     def helper(self, root):
         # edge
         if not root:
            return "#"

         _left = self.helper(root.left)
         _right = self.helper(root.right)

         path = "{}-{}-{}".format(root.val, _left, _right)
         #path.append()
         
         if path not in self.p_map:
            p_map[path] = 1


         # ???
         return path
```
"""



# V0
class Solution(object):
    def longestConsecutive(self, root):
        pass


# V0-1
# IDEA: DFS (gemini)
class Solution(object):
    def longestConsecutive(self, root):
        # Edge case: empty tree
        if not root:
            return 0
            
        self.max_len = 0
        
        # Helper function for DFS traversal
        def dfs(node, parent_val, current_len):
            if not node:
                return
            
            # If the current node is consecutive to its parent, extend the streak
            if node.val == parent_val + 1:
                current_len += 1
            else:
                # Otherwise, the streak breaks. Reset to 1.
                current_len = 1
                
            # Update the global maximum length
            self.max_len = max(self.max_len, current_len)
            
            # Traverse left and right, passing down the current node's value and the streak
            dfs(node.left, node.val, current_len)
            dfs(node.right, node.val, current_len)
            
        # Start the DFS. We pass (root.val - 1) so the root always counts as length 1
        dfs(root, root.val - 1, 0)
        
        return self.max_len


# V0-2
# IDEA: DFS (gpt)
class Solution(object):
    def longestConsecutive(self, root):
        if not root:
            return 0

        self.max_len = 1

        def helper(node, prev_val, curr_len):
            if not node:
                return

            # Continue the consecutive sequence
            if node.val == prev_val + 1:
                curr_len += 1
            else:
                curr_len = 1

            self.max_len = max(self.max_len, curr_len)

            helper(node.left, node.val, curr_len)
            helper(node.right, node.val, curr_len)

        helper(root, root.val - 1, 0)

        return self.max_len


# V0
# IDEA : DFS
# time = O(n), each node visited once
# space = O(h), h = tree height (recursion stack)
class Solution(object):
    def longestConsecutive(self, root):
        if not root:
            return 0

        self.result = 0
        self.helper(root, 1)

        return self.result

    def helper(self, root, curLen):
        self.result = curLen if curLen > self.result else self.result
        if root.left:
            if root.left.val == root.val + 1:
                self.helper(root.left, curLen + 1)
            else:
                self.helper(root.left, 1)
        if root.right:
            if root.right.val == root.val + 1:
                self.helper(root.right, curLen + 1)
            else:
                self.helper(root.right, 1)

# V0'
# IDEA : BFS
# time = O(n), each node visited once
# space = O(n), stack can hold up to O(n) nodes
class Solution(object):
    def longestConsecutive(self, root):
        if root is None:
            return 0

        stack = list()
        stack.append((root, 1))
        maxLen = 1
        while len(stack) > 0:
            node, pathLen = stack.pop()
            if node.left is not None:
                if node.val + 1 == node.left.val:
                    stack.append((node.left, pathLen + 1))
                    maxLen = max(maxLen, pathLen + 1)
                else:
                    stack.append((node.left, 1))
            if node.right is not None:
                if node.val + 1 == node.right.val:
                    stack.append((node.right, pathLen + 1))
                    maxLen = max(maxLen, pathLen + 1)
                else:
                    stack.append((node.right, 1))

        return maxLen

# V0''
# IDEA : BFS
# TODO : validate this approach
# BFS
# class Solution(object):
#     def longestConsecutive(self, root):
#         # edge case
#         if not root:
#             return 0
#         cache = []
#         q = [[cache, root]]
#         res = []
#         while q:
#             for i in range(len(q)):
#                 cache, tmp = q.pop(0)
#                 if cache:
#                     if tmp.val = cache[-1] + 1:
#                         cache.append(tmp.val)
#                         res = max(res, len(cache))
#                     else:
#                         res = max(res, len(cache))
#                         cache = []
#                 if tmp.left:
#                     q.append([cache, tmp.left])
#
#                 if tmp.right:
#                     q.append([cache, tmp.right])
#         return res

# V1
# https://www.jianshu.com/p/ebabdeed9bca
# IDEA : DFS
# time = O(n), each node visited once
# space = O(h), h = tree height (recursion stack)
class Solution(object):
    def longestConsecutive(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        self.result = 0
        self.helper(root, 1)

        return self.result

    def helper(self, root, curLen):
        self.result = curLen if curLen > self.result else self.result
        if root.left:
            if root.left.val == root.val + 1:
                self.helper(root.left, curLen + 1)
            else:
                self.helper(root.left, 1)
        if root.right:
            if root.right.val == root.val + 1:
                self.helper(root.right, curLen + 1)
            else:
                self.helper(root.right, 1)

### Test case : dev

# V1'
# IDEA : BFS
# https://baihuqian.github.io/2018-08-01-binary-tree-longest-consecutive-sequence/
# time = O(n), each node visited once
# space = O(n), stack can hold up to O(n) nodes
class Solution(object):
    def longestConsecutive(self, root):
        if root is None:
            return 0

        stack = list()
        stack.append((root, 1))
        maxLen = 1
        while len(stack) > 0:
            node, pathLen = stack.pop()
            if node.left is not None:
                if node.val + 1 == node.left.val:
                    stack.append((node.left, pathLen + 1))
                    maxLen = max(maxLen, pathLen + 1)
                else:
                    stack.append((node.left, 1))
            if node.right is not None:
                if node.val + 1 == node.right.val:
                    stack.append((node.right, pathLen + 1))
                    maxLen = max(maxLen, pathLen + 1)
                else:
                    stack.append((node.right, 1))

        return maxLen

# V1'
# https://eugenejw.github.io/2017/08/leetcode-298
# IDEA : DFS
# time = O(n), each node visited once
# space = O(h), h = tree height (recursion stack)
class Solution(object):
    def __init__(self):
        self.ret = 0
        
    def longestConsecutive(self, root):
        """
        :type root: TreeNode
        :rtype:     int
        :algorithm: Pre-order DFS, O(n)
        :runtimr:   188ms
        """
        if not root:
            return 0
        self.dfs(root, 1)
        return self.ret
        
    def dfs(self, root, carry):
        """
        :type root: TreeNode
        :rtype: void
        """
        self.ret = max(self.ret, carry)
        if root.left:
            if root.left.val - 1 == root.val:
                self.dfs(root.left, carry+1)
            else:
                self.dfs(root.left, 1)

        if root.right:
            if root.right.val - 1 == root.val:
                self.dfs(root.right, carry+1)
            else:
                self.dfs(root.right, 1)

# V1
# https://blog.csdn.net/qq508618087/article/details/50883425
# JAVA
# /**
#  * Definition for a binary tree node.
#  * struct TreeNode {
#  *     int val;
#  *     TreeNode *left;
#  *     TreeNode *right;
#  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
#  * };
#  */
# class Solution {
# public:
#     void DFS(TreeNode* root, int pre, int len, int& Max)
#     {
#         if(!root) return;
#         len = (root->val==pre+1)?len+1:1;
#         Max = max(Max, len);
#         DFS(root->left, root->val, len, Max);
#         DFS(root->right, root->val, len, Max);
#     }
#    
#     int longestConsecutive(TreeNode* root) {
#         if(!root) return 0;
#         int Max = 1;
#         DFS(root, root->val, 0, Max);
#         return Max;
#     }
# };

# V2
# time = O(n)
# space = O(h), h = tree height (recursion stack)
class Solution(object):
    def longestConsecutive(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.max_len = 0

        def longestConsecutiveHelper(root):
            if not root:
                return 0

            left_len = longestConsecutiveHelper(root.left)
            right_len = longestConsecutiveHelper(root.right)

            cur_len = 1
            if root.left and root.left.val == root.val + 1:
                cur_len = max(cur_len, left_len + 1)
            if root.right and root.right.val == root.val + 1:
                cur_len = max(cur_len, right_len + 1)

            self.max_len = max(self.max_len, cur_len)

            return cur_len

        longestConsecutiveHelper(root)
        return self.max_len

File: leetcode_java/src/main/java/LeetCodeJava/Recursion/BinaryTreeLongestConsecutiveSequence.java
package LeetCodeJava.Recursion;

// https://leetcode.com/problems/binary-tree-longest-consecutive-sequence/description/
// https://leetcode.ca/all/298.html

import LeetCodeJava.DataStructure.TreeNode;

import java.util.*;

/**
 *
 *  298. Binary Tree Longest Consecutive Sequence
 * Given a binary tree, find the length of the longest consecutive sequence path.
 *
 * The path refers to any sequence of nodes from some starting node to any node in the tree along the parent-child connections. The longest consecutive path need to be from parent to child (cannot be the reverse).
 *
 * Example 1:
 *
 * Input:
 *
 *    1
 *     \
 *      3
 *     / \
 *    2   4
 *         \
 *          5
 *
 * Output: 3
 *
 * Explanation: Longest consecutive sequence path is 3-4-5, so return 3.
 * Example 2:
 *
 * Input:
 *
 *    2
 *     \
 *      3
 *     /
 *    2
 *   /
 *  1
 *
 * Output: 2
 *
 * Explanation: Longest consecutive sequence path is 2-3, not 3-2-1, so return 2.
 * Difficulty:
 * Medium
 * Lock:
 * Prime
 * Company:
 * Amazon Facebook Google Uber
 *
 */
public class BinaryTreeLongestConsecutiveSequence {

    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    // V0
    // IDEA : DFS + BFS (fix by gpt)
    // TODO : validate it
    int longestCnt = 0;

    public int longestConsecutive(TreeNode root) {
        if (root == null) {
            return 0;
        }

        // bfs
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            TreeNode node = q.poll();  // Poll the node from the queue
            int length = dfsCount(node);  // Use the node for dfs
            longestCnt = Math.max(longestCnt, length);

            if (node.left != null) {
                q.add(node.left);
            }
            if (node.right != null) {
                q.add(node.right);
            }
        }

        return longestCnt;
    }

    private int dfsCount(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int leftCount = 0, rightCount = 0;

        if (root.left != null && root.val + 1 == root.left.val) {
            leftCount = 1 + dfsCount(root.left);  // Count consecutive path on the left
        }

        if (root.right != null && root.val + 1 == root.right.val) {
            rightCount = 1 + dfsCount(root.right);  // Count consecutive path on the right
        }

        /**
         * NOTE !!!
         *
         *  simply compare leftCount and rightCount,
         *  no need to do below :
         *
         *
         *  return Math.max(
         *                 Math.max(this.dfsCount(root.left), this.dfsCount(root.right)),
         *                 this.dfsCount(root)
         *         );
         *
         * above code may have issue below :
         *
         * 	1.	Redundant Call: this.dfsCount(root) is included in the comparison with dfsCount(root.left) and dfsCount(root.right), leading to redundant recursion. This will likely cause an infinite loop or unnecessary repeated calculations since dfsCount(root) will keep calling itself.
         * 	2.	Logic Error: The comparison includes the root itself in a way that doesn’t make sense in the context of a consecutive sequence. You want to evaluate the left and right children, not compare the root node against itself.
         * 	3.	Performance Degradation: Since this.dfsCount(root) is unnecessarily called multiple times, it can dramatically increase the recursion depth and the overall computational complexity.
         *
         */
        // Return the max length between left and right consecutive paths
        return Math.max(leftCount, rightCount) + 1;  // Include current node
    }

    // V0-0-1
    // IDEA: DFS + PATH SUM + CONSECUTIVE PATH CHECK (fixed by gpt)
    // LC 298 - Binary Tree Longest Consecutive Sequence
// DFS + pathMap2 (node -> path from root to this node) + post traversal to get max consecutive path
    Map<TreeNode, List<Integer>> pathMap2 = new HashMap<>();

    public int longestConsecutive_0_0_1(TreeNode root) {
        // edge
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }

        // Build pathMap2 using DFS
        getNodePath(root, new ArrayList<>());

        // Iterate over all stored paths and find max consecutive length
        int maxConsecutivePathLen = 0;
        for (List<Integer> path : pathMap2.values()) {
            if (path.size() == 0) continue;
            if (isConsecutive(path)) {
                maxConsecutivePathLen = Math.max(maxConsecutivePathLen, path.size());
            }
        }
        return maxConsecutivePathLen;
    }

    private void getNodePath(TreeNode node, List<Integer> path) {
        if (node == null) {
            return;
        }

        // Add current node to path
        path.add(node.val);

        // ⚠️ Important: store a COPY of the current path for this node
        pathMap2.put(node, new ArrayList<>(path));

        // Recurse left and right
        getNodePath(node.left, path);
        getNodePath(node.right, path);

        // Backtrack (undo)
        path.remove(path.size() - 1);
    }

    private boolean isConsecutive(List<Integer> path) {
        int prev = path.get(0);
        for (int i = 1; i < path.size(); i++) {
            if (path.get(i) != prev + 1) {
                return false;
            }
            prev = path.get(i);
        }
        return true;
    }

    // V0-1
    // IDEA: DFS (fixed by gpt)
    // TODO: validate
    public int longestConsecutive_0_1(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return dfs(root, root.val - 1, 0);
    }

    private int dfs(TreeNode node, int prevVal, int currLen) {
        if (node == null) {
            return currLen;
        }

        // Check consecutive condition
        if (node.val == prevVal + 1) {
            currLen = currLen + 1;
        } else {
            currLen = 1;
        }

        int leftMax = dfs(node.left, node.val, currLen);
        int rightMax = dfs(node.right, node.val, currLen);

        // Return max among current, left subtree, right subtree
        return Math.max(currLen, Math.max(leftMax, rightMax));
    }

    // V0-2
    // IDEA: DFS (gemini)
    // TODO: validate
    // Global variable to track the maximum consecutive path length found anywhere in the tree.
    private int maxLen = 0;

    public int longestConsecutive_0_2(TreeNode root) {
        if (root == null) {
            return 0;
        }

        // Start the DFS. The return value of the helper is discarded in the main function;
        // the global maxLen is updated inside the helper.
        dfs(root);

        return maxLen;
    }

    /**
     * Finds the length of the longest consecutive sequence *ending* at the current node.
     * @param node The current node.
     * @return The length of the longest consecutive sequence ending at 'node'.
     */
    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        // Recursively find the longest consecutive path lengths ending at children.
        int leftLen = dfs(node.left);
        int rightLen = dfs(node.right);

        // Current length starts at 1 (the node itself).
        int currentLen = 1;

        // Check if the current node extends the left path.
        if (node.left != null && node.left.val == node.val + 1) {
            currentLen = Math.max(currentLen, leftLen + 1);
        }

        // Check if the current node extends the right path.
        if (node.right != null && node.right.val == node.val + 1) {
            currentLen = Math.max(currentLen, rightLen + 1);
        }

        // Update the global maximum path length.
        maxLen = Math.max(maxLen, currentLen);

        // Return the length of the path that can be EXTENDED by the parent.
        // The parent can only choose to extend the current node's path in ONE direction (upwards).
        return currentLen;
    }


    // V1
    // IDEA : TREE, RECURSION
    // https://leetcode.ca/2016-09-23-298-Binary-Tree-Longest-Consecutive-Sequence/
    private int ans;

    public int longestConsecutive_1_1(TreeNode root) {
        dfs_1(root);
        return ans;
    }

    private int dfs_1(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int l = dfs_1(root.left) + 1;
        int r = dfs_1(root.right) + 1;
        if (root.left != null && root.left.val - root.val != 1) {
            l = 1;
        }
        if (root.right != null && root.right.val - root.val != 1) {
            r = 1;
        }
        int t = Math.max(l, r);
        ans = Math.max(ans, t);
        return t;
    }

    // V2
    // IDEA : DFS
    // https://walkccc.me/LeetCode/problems/298/#__tabbed_1_2
    public int longestConsecutive_2_1(TreeNode root) {
        if (root == null)
            return 0;
        return dfs_2(root, -1, 0, 1);
    }

    private int dfs_2(TreeNode root, int target, int length, int mx) {
        if (root == null)
            return mx;
        if (root.val == target)
            mx = Math.max(mx, ++length);
        else
            length = 1;
        return Math.max(dfs_2(root.left, root.val + 1, length, mx),
                dfs_2(root.right, root.val + 1, length, mx));
    }

    // V3
    // IDEA : DFS
    // https://blog.csdn.net/qq_46105170/article/details/108481001
    private int res;

    public int longestConsecutive_3(TreeNode root) {
        dfs_3(root);
        return res;
    }

    // The dfs function starts at the root and returns the length of the longest consecutive sequence
    private int dfs_3(TreeNode root) {
        if (root == null) {
            return 0;
        }

        // Initialize the length to 1
        int len = 1;

        // Get the longest consecutive path from left and right child nodes
        int left = dfs_3(root.left), right = dfs_3(root.right);

        // If the left child is consecutive, increase the length
        if (root.left != null && root.left.val == root.val + 1) {
            len = Math.max(len, 1 + left);
        }

        // If the right child is consecutive, increase the length
        if (root.right != null && root.right.val == root.val + 1) {
            len = Math.max(len, 1 + right);
        }

        // Update the global result
        res = Math.max(res, len);

        // Return the length of the longest consecutive path starting from the current node
        return len;
    }



}
